### ISSUE TITLE
Error: Can't resolve 'persisted_queries.json'

### ISSUE THREAD

**Issue by andylhansen** [sysgears/persistgraphql-webpack-plugin#1]
I am having trouble getting my webpack configuration to work. I get the error: Error: Can't resolve 'persisted_queries.json'. Webpack still outputs the persisted_queries.json file to the filesystem, but it doesn't let me include it in my app's code. I'm using webpack 2.4.1.

This is my webpack configuration file.
```
const { resolve, join } = require('path');
const webpack = require('webpack');
const path = require('path');
const fs = require('fs');
const HtmlWebpackPlugin = require('html-webpack-plugin');
var PersistGraphQLPlugin = require('persistgraphql-webpack-plugin');
const cheerio = require('cheerio');

const dllPath = path.resolve(process.cwd(), 'dlls');

module.exports = {
  entry: [
    'react-hot-loader/patch',
    // activate HMR for React

    'webpack-hot-middleware/client?path=/__webpack_hmr&timeout=20000',
    // bundle the client for webpack-dev-server
    // and connect to the provided endpoint

    path.join(process.cwd(), 'app/app.js')
    // the entry point of our app
  ],

  output: {
    filename: '[name].js',
    chunkFilename: '[name].chunk.js',

    path: join(process.cwd(), 'public'),

    publicPath: '/'
    // necessary for HMR to know where to load the hot update chunks
  },

  devtool: 'none',

  module: {
    rules: [
      {
        test: /\.jsx?$/,
        use: [
          {
            loader: 'babel-loader',
            query: {
              babelrc: false,
              presets: [
                'flow',
                ['es2015', { modules: false }],
                'stage-0',
                'react',
              ],
              plugins: [
                "styled-components",
                "react-hot-loader/babel"
              ]
            },
          }, {
            loader: 'persistgraphql-webpack-plugin/js-loader',
          }
        ],
        exclude: /node_modules/
      }, {
        test: /\.(graphql|gql)$/,
        exclude: /node_modules/,
        loaders: [
          'graphql-tag/loader',
          'persistgraphql-webpack-plugin/graphql-loader',
        ],
      },
    ],
  },

  plugins: [

    new webpack.HotModuleReplacementPlugin(),
    // enable HMR globally

    new webpack.NamedModulesPlugin(),
    // prints more readable module names in the browser console on HMR updates

    new HtmlWebpackPlugin({
      inject: true, // Inject all files that are generated by webpack, e.g. bundle.js
      templateContent: templateContent(),
    }),

    new webpack.DllReferencePlugin({
      context: '.',
      manifest: require(path.resolve(dllPath, 'vendor.json')),
    }),

    new PersistGraphQLPlugin({filename: 'persisted_queries.json'}),

  ],
  resolve: {
    modules: ['node_modules', 'app'],
  }
};


function templateContent() {
  const html = fs.readFileSync(
    path.resolve(process.cwd(), 'app/index.html')
  ).toString();

  const doc = cheerio(html);
  const body = doc.find('body');

  body.append('<script data-dll="true" src="/vendor.dll.js"></script>');

  return doc.toString();
}
```
In my client-side app, I have a file called client.js. This is where I try to include the persisted_queries.json file.
```
import { ApolloClient } from 'react-apollo';
import { createBatchingNetworkInterface } from 'apollo-client';
import { addPersistedQueries } from 'persistgraphql';

let networkInterface = createBatchingNetworkInterface({
  opts: {
    credentials: "same-origin",
  },
  batchInterval: 20,
  uri: "/graphql",
});

const queryMap = require('persisted_queries.json');
networkInterface = addPersistedQueries(networkInterface, queryMap);

const client = new ApolloClient(networkInterface);

export default client;
```

Is there something I've missed in my configuration that would cause this error?

**Comment by larixer** [comment]
@andylhansen I see this problem with `persistgraphql-webpack-plugin` version `0.1.8`, but not with version `0.1.9`. Could you check if version `0.1.9` (the latest at the moment) works for you?

**Comment by benjie** [comment]
I am getting this issue with 0.1.9

```js
// ...
const PersistGraphQLPlugin = require('persistgraphql-webpack-plugin');
const persistGraphQLPlugin = new PersistGraphQLPlugin({ filename: 'persisted_queries.json' });
// ...
const config = {};

config.plugins = [
  //...
  new LodashModuleReplacementPlugin(),
  persistGraphQLPlugin,
];

config.context = clientDir;

config.resolve = {
  modules: [
    clientDir,
    nodeModulesDir,
  ],
};

config.entry = {
  app: ['./index.js'],
};

config.output = {
  path: distDir,
  filename: dev ? '[name].js' : '[name].[chunkhash:10].js',
  chunkFilename: dev ? '[name].js' : '[name].[chunkhash:10].js',
  publicPath: '/',
};

config.module = {
  rules: [{
    test: /\.js$/,
    exclude: /node_modules/,
    use: [{
      loader: 'babel-loader',
      options: {
        babelrc: false,
        presets: [
          'react',
          ['es2015', {
            loose: false,
            modules: false
          }],
          'stage-0'
        ],
        plugins: [
          'lodash',
          'transform-runtime',
          'react-require'
        ],
      }
    }, {
      loader: 'persistgraphql-webpack-plugin/js-loader',
    }]
  }, {
    test: /\.css$/,
    include: /node_modules/,
    use: [
      'style-loader',
      'css-loader'
    ],
  }, {
    test: /\.(graphql|gql)$/,
    exclude: /node_modules/,
    use: [
      'graphql-tag/loader',
      'persistgraphql-webpack-plugin/graphql-loader',
    ],
  }]
};

if (dev) {
  config.module.rules.concat({
    test: /\.js$/,
    exclude: /node_modules/,
    use: [{
      loaders: [
        'react-hot-loader/webpack',
      ],
    }],
  });

  config.entry.app.unshift(
    'react-hot-loader/patch',
    'webpack-hot-middleware/client?path=/__webpack_hmr&reload=true'
  );

  config.plugins.push(
    new webpack.HotModuleReplacementPlugin(),
    new webpack.NoEmitOnErrorsPlugin()
  );
} else {
// ... 
}

module.exports = config;
```

and

```js
import { ApolloClient, createNetworkInterface } from 'react-apollo';
import persistgraphql from 'persistgraphql';
import queryMap from 'persisted_queries.json';

const networkInterface = createNetworkInterface({
  uri: '/graphql',
  opts: {
    credentials: 'same-origin',
  },
});

persistgraphql.addPersistedQueries(networkInterface, queryMap);

const client = new ApolloClient({
  networkInterface,
  addTypename: true,
  dataIdFromObject: (result) => {
    if (result.id && result.__typename) {
      return result.__typename + result.id;
    }
    return null;
  }
});

export default client;
```

**Comment by benjie** [comment]
Webpack 2.3.2 btw.

**Comment by larixer** [comment]
@benjie Could you set `DEBUG=webpack-virtual-modules` environment variable when starting webpack and paste webpack `console` and Error stack trace here?

**Comment by larixer** [comment]
@benjie Also you might want to consider setting `moduleName` option of this plugin to an absolute path to `persisted_queries.json`, e.g.:
``` js
new PersistGraphQLPlugin({ moduleName: path.resolve('node_modules/persisted_queries.json') });
```

**Comment by benjie** [comment]
```
Mon, 24 Apr 2017 15:51:23 GMT webpack-virtual-modules undefined Write module: /Users/benjiegillam/Documents/CLIENT_NAME/client/node_modules/persisted_queries.json {"mutat...
```
Interesting; my actual node_modules folder is one folder up from that; but since I'm importing from `client/apollo.js` I'd expect it to work nonetheless...

```
ERROR in ./client/apollo.js
Module not found: Error: Can't resolve 'persisted_queries.json' in '/Users/benjiegillam/Documents/CLIENT_NAME/client'
 @ ./client/apollo.js 3:0-46
```

**Comment by larixer** [comment]
@benjie I think `moduleName` should be required option for this plugin, to not confuse people and let them specify exact virtual module location. I don't know yet how to implement proper location guessing inside this plugin.

**Comment by larixer** [comment]
@benjie @andylhansen I have made `moduleName` a required option for this plugin and published this package as version `0.2.0` on npm. Hopefully this will reduce confusion about location of generated virtual webpack module

**Comment by benjie** [comment]
Confirmed, the following no longer produces that error:

```js
const persistGraphQLPlugin = new PersistGraphQLPlugin({
  filename: 'persisted_queries.json',
  moduleName: path.resolve(`${__dirname}/../node_modules/persisted_queries.json`),
});
```

**Comment by benjie** [comment]
I just hit this again (my fault) in a monorepo which I just moved over to use yarn workspace. The issue was that the `node_modules` folder in the package no longer needed to exist, so on CI it was failing (whereas locally the folder still existed so it was fine). Took a while to figure out, so adding this note for anyone else who hits this in future. Solution I used was simply to `mkdir -p node_modules` as part of the build; but another alternative would be using the `moduleName` to point to `../../node_modules` - i.e. the monorepo root.
